core/finality-grandpa/src/communication/periodic.rs

```diff
@@ -106,3 +114,141 @@ pub(super) fn neighbor_packet_worker<B, N>(net: N) -> (
 
 	(work, NeighborPacketSender(tx))
 }
+
+/// A background worker for performing block announcements.
+struct BlockAnnouncer<B: BlockT, N> {
+	net: N,
+	block_rx: mpsc::UnboundedReceiver<B::Hash>,
+	latest_voted_blocks: VecDeque<B::Hash>,
+	reannounce_after: Duration,
+	delay: Delay,
+}
+
+/// A background worker for announcing block hashes to peers. The worker keeps
+/// track of `LATEST_VOTED_BLOCKS_TO_ANNOUNCE` and periodically announces these
+/// blocks to all peers if no new blocks to announce are noted (i.e. presumably
+/// GRANDPA progress is stalled).
+pub(super) fn block_announce_worker<B: BlockT, N: Network<B>>(net: N) -> (
+	impl Future<Item = (), Error = ()>,
+	BlockAnnounceSender<B>,
+) {
+	block_announce_worker_aux(net, REBROADCAST_AFTER)
+}
+
+#[cfg(test)]
+pub(super) fn block_announce_worker_with_delay<B: BlockT, N: Network<B>>(
+	net: N,
+	reannounce_after: Duration,
+) -> (
+	impl Future<Item = (), Error = ()>,
+	BlockAnnounceSender<B>,
+) {
+	block_announce_worker_aux(net, reannounce_after)
+}
+
+fn block_announce_worker_aux<B: BlockT, N: Network<B>>(
+	net: N,
+	reannounce_after: Duration,
+) -> (
+	impl Future<Item = (), Error = ()>,
+	BlockAnnounceSender<B>,
+) {
+	let latest_voted_blocks = VecDeque::with_capacity(LATEST_VOTED_BLOCKS_TO_ANNOUNCE);
+
+	let (block_tx, block_rx) = mpsc::unbounded();
+
+	let announcer = BlockAnnouncer {
+		net,
+		block_rx,
+		latest_voted_blocks,
+		reannounce_after,
+		delay: Delay::new(Instant::now() + reannounce_after),
+	};
+
+	(announcer, BlockAnnounceSender(block_tx))
+}
+
+
+impl<B: BlockT, N> BlockAnnouncer<B, N> {
+	fn note_block(&mut self, block: B::Hash) -> bool {
+		if !self.latest_voted_blocks.contains(&block) {
+			if self.latest_voted_blocks.len() >= LATEST_VOTED_BLOCKS_TO_ANNOUNCE {
+				self.latest_voted_blocks.pop_front();
+			}
+
+			self.latest_voted_blocks.push_back(block);
+
+			return true;
+		}
+
+		false
+	}
+
+	fn reset_delay(&mut self) {
+		self.delay.reset(Instant::now() + self.reannounce_after);
+	}
+}
+
+impl<B: BlockT, N: Network<B>> Future for BlockAnnouncer<B, N> {
+	type Item = ();
+	type Error = ();
+
+	fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+		// note any new blocks to announce and announce them
+		loop {
+			match self.block_rx.poll().expect("unbounded receivers do not error; qed") {
+				Async::Ready(None) => return Ok(Async::Ready(())),
+				Async::Ready(Some(block)) => {
+					if self.note_block(block) {
+						self.net.announce(block);
+						self.reset_delay();
+					}
+				},
+				Async::NotReady => break,
+			}
+		}
+
+		// after the delay fires announce all blocks that we have stored. note
+		// that this only happens if we don't receive any new blocks above for
```

Comment:
This comment is confusing.  The code seems correct, but we will hit this loop even if the timer has not expired.